### PR TITLE
feat(Core/BG): track who accept to join BG

### PR DESCRIPTION
**Pull request by Helias**
Why this? because I would track who joins BGs.
**Scenario**: the same person joins BG with 2 characters and with one of the two do not enter to BG it makes an unbalanced game inside the BG.

Actually, we track only the char that does not enter, but even who enter can be considered an abuser because in this scenario he is the same person with same IP probably.

I will update the tool [BG-Queue-Abuser-Viewer](https://github.com/Helias/BG-Queue-Abuser-Viewer) for this.

## Changes Proposed:
-  track the players who accepts to join BG and not only who leave it

## Tests Performed:
- tested in-game

## How to Test the Changes:
- .debug bg
- join BG
- enter to BG and check battleground_deserters table, there will be a type 5 that stands for BG accepted

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Handlers/BattleGroundHandler.cpp [hunk 1]
```diff
@@ -463,6 +463,14 @@ void WorldSession::HandleBattleFieldPortOpcode(WorldPacket& recvData)
 
                 _player->SetBattlegroundId(bg->GetInstanceID(), bg->GetBgTypeID(), queueSlot, true, bgTypeId == BATTLEGROUND_RB, teamId);
 
+                if (sWorld->getBoolConfig(CONFIG_BATTLEGROUND_TRACK_DESERTERS))
+                {
+                    PreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_INS_DESERTER_TRACK);
+                    stmt->setUInt32(0, _player->GetGUIDLow());
+                    stmt->setUInt8(1, BG_DESERTION_TYPE_NONE);
+                    CharacterDatabase.Execute(stmt);
+                }
```
Comment (FrancescoBorzi) [inline src/server/game/Handlers/BattleGroundHandler.cpp:472]: feels a bit wrong.. we are tracking who does NOT desert inside a `TRACK_DESERTERS` feature. Why not simply adding a new hook such as `OnPlayerJoinedBG` ?

Comment (Helias) [reply src/server/game/Handlers/BattleGroundHandler.cpp]: done

File: src/server/game/Battlegrounds/Battleground.h [hunk 2]
```diff
@@ -40,6 +40,7 @@ enum BattlegroundDesertionType
     BG_DESERTION_TYPE_LEAVE_QUEUE       = 2, // player is invited to join and refuses to do it
     BG_DESERTION_TYPE_NO_ENTER_BUTTON   = 3, // player is invited to join and do nothing (time expires)
     BG_DESERTION_TYPE_INVITE_LOGOUT     = 4, // player is invited to join and logs out
+    BG_DESERTION_TYPE_NONE              = 5, // player is invited to join and accepted to do it
```
Comment (FrancescoBorzi) [inline src/server/game/Battlegrounds/Battleground.h:43]: remove this

File: src/server/game/Handlers/BattleGroundHandler.cpp [hunk 3]
```diff
@@ -464,6 +464,8 @@ void WorldSession::HandleBattleFieldPortOpcode(WorldPacket& recvData)
                 _player->SetBattlegroundId(bg->GetInstanceID(), bg->GetBgTypeID(), queueSlot, true, bgTypeId == BATTLEGROUND_RB, teamId);
 
                 sBattlegroundMgr->SendToBattleground(_player, ginfo.IsInvitedToBGInstanceGUID, bgTypeId);
+
+                sScriptMgr->OnPlayerJoinedBattleground(_player);
```
Comment (FrancescoBorzi) [inline src/server/game/Handlers/BattleGroundHandler.cpp:468]: what about sending more info?? such as the BG type

File: src/server/game/Battlegrounds/Battleground.h (hunk 2 above)
Comment (Helias) [reply src/server/game/Battlegrounds/Battleground.h]: done

File: src/server/game/Handlers/BattleGroundHandler.cpp (hunk 3 above)
Comment (Helias) [reply src/server/game/Handlers/BattleGroundHandler.cpp:468]: like?

Comment (FrancescoBorzi) [reply src/server/game/Handlers/BattleGroundHandler.cpp:468]: `bgTypeId` for example

**Comment by Hacki95** [comment]
Merge conflicts

**Pull request closed without merging**